test/mail-test-srv/main.go

```diff
@@ -0,0 +1,225 @@
+		return fmt.Errorf("readline: %v", err)
+	}
+	if string(line) != expected {
+		return fmt.Errorf("Expected %s, got %s", expected, line)
+	}
+	return nil
+}
+
+func getLine(buf *bufio.Reader) (string, error) {
+	line, _, err := buf.ReadLine()
+	if err != nil {
+		return "", fmt.Errorf("readline: %v", err)
+	}
+	return string(line), nil
+}
+
+var mailFromRegex = regexp.MustCompile("^MAIL FROM:<(.*)>\\s*BODY=8BITMIME\\s*$")
+var rcptToRegex = regexp.MustCompile("^RCPT TO:<(.*)>\\s*$")
+var dataRegex = regexp.MustCompile("^DATA\\s*$")
+var smtpErr501 = []byte("501 syntax error in parameters or arguments \r\n")
+var smtpOk250 = []byte("250 OK \r\n")
+var authLine string
+
+func handleConn(conn net.Conn) {
+	defer conn.Close()
+	auditlogger := blog.GetAuditLogger()
+	auditlogger.Info(fmt.Sprintf("mailcatch: Got connection from %s", conn.RemoteAddr()))
+
+	readBuf := bufio.NewReader(conn)
+	conn.Write([]byte("220 smtp.example.com ESMTP\r\n"))
+	if err := expectLine(readBuf, "EHLO localhost"); err != nil {
+		log.Printf("mailcatch: %s: %v\n", conn.RemoteAddr(), err)
+		return
+	}
+	conn.Write([]byte("250-PIPELINING\r\n"))
+	conn.Write([]byte("250-AUTH PLAIN LOGIN\r\n"))
+	conn.Write([]byte("250 8BITMIME\r\n"))
+	if err := expectLine(readBuf, authLine); err != nil {
+		log.Printf("mailcatch: %s: %v\n", conn.RemoteAddr(), err)
+		return
+	}
+	conn.Write([]byte("235 2.7.0 Authentication successful\r\n"))
+	auditlogger.Info(fmt.Sprintf("mailcatch: Successful auth from %s", conn.RemoteAddr()))
+	// == END authentication
+
+	// necessary commands:
+	// MAIL RCPT DATA QUIT
+
+	var fromAddr string
+	var toAddr []string
+	var msgBuf bytes.Buffer
+
+	clearState := func() {
+		fromAddr = ""
+		toAddr = nil
+		msgBuf = bytes.Buffer{}
+	}
+
+	var quit = false
+	for !quit {
+		line, err := getLine(readBuf)
+		if err != nil {
+			log.Printf("mailcatch: %s: getline: %v\n", conn.RemoteAddr(), err)
+			return
+		}
+
+		cmdSplit := strings.SplitN(line, " ", 2)
+		cmd := cmdSplit[0]
+		switch cmd {
+		case "QUIT":
+			quit = true
+			conn.Write([]byte("221 Bye \r\n"))
+			break
+		case "RSET":
+			clearState()
+			conn.Write(smtpOk250)
+		case "NOOP":
+			conn.Write(smtpOk250)
+		case "MAIL":
+			clearState()
+			matches := mailFromRegex.FindStringSubmatch(line)
+			if matches == nil {
+				log.Printf("mailcatch: %s: MAIL FROM parse error\n", conn.RemoteAddr())
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
+				continue
+			}
+			fromAddr = addr.Address
+			conn.Write(smtpOk250)
+		case "RCPT":
+			matches := rcptToRegex.FindStringSubmatch(line)
+			if matches == nil {
+				conn.Write(smtpErr501)
+				continue
+			}
+			addr, err := mail.ParseAddress(matches[1])
+			if err != nil {
+				log.Printf("mailcatch: %s: addr parse error: %v\n", conn.RemoteAddr(), err)
+				conn.Write(smtpErr501)
```

Comment:
I think it's fine to just panic here. And more desirable since it may catch problems that would otherwise go unnoticed.